### PR TITLE
Implemented addEvent for plugin development.

### PR DESCRIPTION
**Pull request by kiatng**
`mwrun` is broken because the method `addEvent` is  unfinished:

https://github.com/ShipStream/middleware/blob/21d7d52c66d65901d52c77dcec5a4956c1e84154/lib/Plugin/Abstract.php#L435-L438

### CODE REVIEW

**Review by colinmollenhour — COMMENTED** [review]
Oops!

Rather than blocking for the delay time please have it do the following:
- Add the event data to a private array like `$this->_eventQueue[] = [$method, $data, $delayTime];`
- When the execution is finished run all of the queued events.

This will better simulate the background processing that happens in production by deferring these events until the end of the action. E.g. if 10 orders are found then the 10 events will be queued instantly and then each one imported rather than importing all 10 in the "foreground".

**Comment by kiatng** [comment]
Like this?

```php
    protected $_eventQueue = []; // Fake queue to simulate a table in database.

    final public function addEvent($method, array $data, $delayTime = NULL)
    {
        if (is_int($delayTime)) {
            $this->_eventQueue[] = [$method, $data, $delayTime]; // Add entry to table in non-simulation.
        } else {
           $this->$method(new Varien_Object($data));
       }
       // Queue simulation. Use cron in non-simulation.
       foreach ($this->_eventQueue as $queue) {
           list($method, $data) = $queue;
           $this->$method(new Varien_Object($data));
       }
    }
```

**Comment by colinmollenhour** [comment]
Sorta. The `// Queue simulation.` part needs to be moved to somewhere after the main execution has completed such as https://github.com/ShipStream/middleware/blob/f39c7856941a88d0c549984d921808769bf66f2c/app/bootstrap.php#L71. This may require the events array to be stored in the Middleware instance rather than the plugin instance to avoid adding new public methods to the latter.

**Comment by kiatng** [comment]
I hope I got it right this time.

File: lib/Plugin/Abstract.php [hunk 1]
```diff
@@ -442,7 +442,11 @@ final public function resolveError($rawData)
      */
     final public function addEvent($method, array $data, $delayTime = NULL)
     {
-        // TODO
+        if (is_int($delayTime)) {
+            $this->middleware->addEventQueue([$method, new Varien_Object($data), $delayTime]);
+        } else {
+           $this->$method(new Varien_Object($data));
```
Comment (colinmollenhour) [inline lib/Plugin/Abstract.php]: Very good! But, to most accurately represent what happens on production the event should be added to the queue even when there is no `$delayTime`.

Comment (kiatng) [reply lib/Plugin/Abstract.php]: Makes sense now.

**Comment by kiatng** [comment]
@colinmollenhour Please merge if it's good.

**Comment by colinmollenhour** [comment]
Looks good!

**Pull request merged**